contracts/products/SoteriaCoverageProduct.sol

```diff
@@ -445,59 +393,162 @@ contract SoteriaCoverageProduct is ISoteriaCoverageProduct, ERC721, EIP712, Reen
-     * @param signer_ The signer to add.
+     * @param maxRateNum_ Desired maxRateNum.
     */
-    function addSigner(address signer_) external override onlyGovernance {
-        require(signer_ != address(0x0), "zero address signer");
-        _isAuthorizedSigner[signer_] = true;
-        emit SignerAdded(signer_);
+    function setMaxRateNum(uint256 maxRateNum_) external override onlyGovernance {
+        _maxRateNum = maxRateNum_;
+        emit MaxRateNumSet(maxRateNum_);
     }
 
     /**
-     * @notice Removes a signer.
+     * @notice set _maxRateDenom.
      * Can only be called by the current [**governor**](/docs/protocol/governance).
-     * @param signer_ The signer to remove.
+     * @param maxRateDenom_ Desired maxRateDenom.
     */
-    function removeSigner(address signer_) external override onlyGovernance {
-        _isAuthorizedSigner[signer_] = false;
-        emit SignerRemoved(signer_);
+    function setMaxRateDenom(uint256 maxRateDenom_) external override onlyGovernance {
+        _maxRateDenom = maxRateDenom_;
+        emit MaxRateDenomSet(maxRateDenom_);
+    }
+
+    /**
+     * @notice set _chargeCycle.
+     * Can only be called by the current [**governor**](/docs/protocol/governance).
+     * @param chargeCycle_ Desired chargeCycle.
+    */
+    function setChargeCycle(uint256 chargeCycle_) external override onlyGovernance {
+        _chargeCycle = chargeCycle_;
+        emit ChargeCycleSet(chargeCycle_);
+    }
+
+    /**
+     * @notice Enables governance to gift (and remove) 'free' cover to specific addresses.
+     * Can only be called by the current [**governor**](/docs/protocol/governance).
+     * @param policyholder_ The policy holder to set reward points for.
+     * @param rewardPoints_ Desired amount of reward points.
+    */
+    function setRewardPoints(address policyholder_, uint256 rewardPoints_) external override onlyGovernance {
+        _rewardPointsOf[policyholder_] = rewardPoints_;
+        emit RewardPointsSet(policyholder_, rewardPoints_);
+    }  
+
+    /**
+     * @notice Charge premiums for each policy holder.
+     * @param holders_ The policy holders.
+     * @param premiums_ The premium amounts in `wei` per policy holder.
+    */
+    function chargePremiums(address[] calldata holders_, uint256[] calldata premiums_) external payable override onlyGovernance whileUnpaused {
+        uint256 count = holders_.length;
+        require(count == premiums_.length, "length mismatch");
+        require(count <= policyCount(), "policy count exceeded");
+        uint256 amountToPayTreasury = 0;
+
+        for (uint256 i = 0; i < count; i++) {
+            // skip computation if policy inactive
+            if ( !policyStatus(_policyOf[holders_[i]]) ) continue;
+            require(premiums_[i] <= ( _maxRateNum * _chargeCycle * coverLimitOf(policyOf(holders_[i])) ) / _maxRateDenom, "Charging more than promised maximum rate");
```

Comment:
- extracting to the function.(mentioned above)
- lowercase error message

Reply:
I am not sure that should we raise error in this statement.   **require(premiums_[i] <= ( _maxRateNum * _chargeCycle * coverLimitOf(policyOf(holders_[i])) ) / _maxRateDenom, "Charging more than promised maximum rate");**

This situation is caused by us. We need to discuss it.